transport/tlscommon/server_config_test.go

```diff
@@ -0,0 +1,81 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package tlscommon
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v2"
+)
+
+func Test_ServerConfig_Serialization_ClientAuth(t *testing.T) {
+	tests := []struct {
+		name       string
+		cfg        ServerConfig
+		clientAuth TLSClientAuth
+	}{{
+		name: "with ca",
+		cfg: ServerConfig{
+			Certificate: CertificateConfig{
+				Certificate: "/path/to/cert.crt",
+				Key:         "/path/to/cert.key",
+			},
+			CAs: []string{"/path/to/ca.crt"},
+		},
+		clientAuth: TLSClientAuthRequired,
+	}, {
+		name: "no ca",
+		cfg: ServerConfig{
+			Certificate: CertificateConfig{
+				Certificate: "/path/to/cert.crt",
+				Key:         "/path/to/cert.key",
+			},
+		},
+		clientAuth: TLSClientAuthNone,
+	}, {
+		name: "with ca and client auth none",
+		cfg: ServerConfig{
+			Certificate: CertificateConfig{
+				Certificate: "/path/to/cert.crt",
+				Key:         "/path/to/cert.key",
+			},
+			CAs:        []string{"/path/to/ca.crt"},
+			ClientAuth: TLSClientAuthNone,
+		},
+		clientAuth: TLSClientAuthNone, // FIXME test fails, it serializes to required
```

Comment:
This test is failing because of the `omitempty` tag on the ClientAuth attribute of the struct.
`TLSClientAuthNone` corresponds with a zero value, so it will not be serialized if it is set in the struct.

If we remove the `omitempty` tag, the "with ca" unit test will fail (none will serialize, and the `Unpack` method will not insert the required val (current default when a CA is specified).
I can try to change ClientAuth to be a pointer value instead, so there is a difference between no value, and a none (0) val